### DISCUSSION TITLE
Meeting Doc - April 22, 2022

### DISCUSSION THREAD

**Discussion by ranchowang** [nextarch/SIG-Microservice#4] [General]
会议内容：

1. 分享go-zero微服务实践经验 @万俊峰
2. 分享服务治理标准化项目进展 @单家骏

-----------------------------------------------

出席情况：
Hongzhi Wang, Shibai Wei, Ke Su, Bo Zhang, Bin Fan, Guoguang Hu, Jiajun Shan, Chuntao Liao, Shun Song, Junfeng Wan, Qiang Guo, Guangming Luo, Le Zhang, Haotian Zhang

-----------------------------------------------

关于go-zero微服务实践经验的交流：

1、go-zero数据治理如何自动实现数据缓存？

go-zero框架自动识别mysql数据表的索引，在redis里生成缓存，框架负责缓存的更新和淘汰

2、go-zero项目提供了api gateway？

是的，go-zero可以根据proto生成restful接口，实现到grpc的转换

3、go-zero基于k8s endpoints做服务发现，如何实现跨k8s集群的服务发现？

跨k8s集群访问需要经过proxy，和集群内的服务调用方式不同。也可以基于注册中心实现跨k8s集群的服务发现，集群内外的服务调用方式可以保持一致

4、go-zero如何实现自适应负载和弹性伸缩？

为了防止流量突发，导致一个服务的全部节点被压垮，go-zero会采集本节点的负载，让负载保持在80-90%，同时k8s hpa配置cpu达到80%进行扩容。因为k8s扩容需要一定的时间，在这段时间内，go-zero自适应负载可以保障本节点不会被压垮

5、go-zero如何保证服务节点的无损下线？

go-zero框架收到进程结束的信号时，先进行反注册，这样不会再有新请求访问过来，然后在处理完现有请求后再退出。对于长连接的情况，go-zero框架作为主调方，如何发现一个被调方服务节点反注册了，就会断开对应的长连接，不会再发请求过去

-----------------------------------------------

关于服务治理标准化项目的交流：

1、服务治理spec初稿的版本使用1.0是否合适？

家骏基于北极星项目的实现输出服务治理spec初稿，认为当前具备可用的实现，因此采用了1.0版本。但是，这个版本还没有经过大家的共同探讨，采用0.1版本更为适合

2、服务治理标准化是搞一个规范还是搞一个项目？

待补充

3、服务治理spec包含可观测性，和已有观测标准是什么关系？

待补充

4、业界有不少类似的标准实现，如何形成更加广泛的共识？

待补充

**Comment by andrewshan** [comment]
2、服务治理标准化是搞一个规范还是搞一个项目？

标准化项目不仅只提供规范，会提供实现。标准化实现包含三个部分，
第一部分是服务治理平面：治理平面面向用户，通过提供可视化界面支持服务治理功能的管控（包括规则的编辑、治理效果的监控等），并下发规则到数据平面。
第二部分是数据平面：数据平面主要面向框架或应用，提供API接口进行功能的暴露。过程中会接受治理平面下发的规则进行，并执行具体的服务治理逻辑处理。
第三部分是开发框架和数据平面的对接和集成：数据平面可以单独使用，更常见的是集成在开发框架中，通过扩展开发框架的拦截器或接口，实现标准化实现与框架的集成。
北极星（Polarismesh）会在现有的基础上支持服务治理标准化实现，go-zero、GoFrame、CloudWeGO 和 Spring Cloud Tencent 等多个开源社区也会集成标准化实现，为各个社区的用户提供开箱即用的实现。这些开源项目都经过了企业内外部用户的验证，在性能和稳定性上都有保障，可以加速标准化的落地。

**Comment by andrewshan** [comment]
3、服务治理spec包含可观测性，和已有观测标准是什么关系？

当前还没有启动可观测性相关的SPEC的制定，到时启动后会和大家一起定。

**Comment by andrewshan** [comment]
4、业界有不少类似的标准实现，如何形成更加广泛的共识？

https://doc.weixin.qq.com/doc/w3_ABgAJAaAACcEBHl1iQPR0KVOYazIx?scode=AJEAIQdfAAoZXtEVnLABgAJAaAACc
